src/python/pants/source/wrapped_globs.py

```diff
@@ -4,37 +4,45 @@
 import os
 from abc import ABC, abstractmethod
 from hashlib import sha1
+from typing import Callable, Iterable, Iterator, List, Sequence, Tuple, cast
 
 from twitter.common.dirutil.fileset import Fileset
+from typing_extensions import TypedDict
 
 from pants.base.build_environment import get_buildroot
-from pants.engine.fs import EMPTY_SNAPSHOT
+from pants.base.parse_context import ParseContext
+from pants.engine.fs import EMPTY_SNAPSHOT, Snapshot
 from pants.util.dirutil import fast_relpath, fast_relpath_optional
 from pants.util.memo import memoized_property
 
 
+class _GlobsDict(TypedDict, total=True):
+  globs: List[str]
+
+
+class Filespec(_GlobsDict, total=False):
+  """A dict of globs (required) and excludes (optional).
+
+  For example: {'globs': ['list', 'of' , 'strings'], 'exclude': [{'globs' : ... }, ...] }
+
+  The globs are in zglobs format."""
+  exclude: List[_GlobsDict]
+
+
 class FilesetWithSpec(ABC):
   """A set of files that keeps track of how we got it."""
 
   @staticmethod
-  def _no_content(path):
-    raise AssertionError('An empty FilesetWithSpec should never have file content requested.')
-
-  @staticmethod
-  def empty(rel_root):
+  def empty(rel_root: str) -> "EagerFilesetWithSpec":
     """Creates an empty FilesetWithSpec object for the given rel_root."""
     return EagerFilesetWithSpec(rel_root, {'globs': []}, EMPTY_SNAPSHOT)
 
   @abstractmethod
-  def matches(self, path_from_buildroot):
-    """
-    Takes in any relative path from build root, and return whether it belongs to this filespec
-
-    :param path_from_buildroot: path relative to build root
-    :return: True if the path matches, else False.
-    """
+  def matches(self, path_from_buildroot: str) -> bool:
+    """Takes in any relative path from build root, and return whether it belongs to this
+    filespec."""
```

Comment:
I think this fits on one line

Reply:
It's something like 103 characters :/

Reply:
Is 120 not the limit?

Reply:
100. Although I would love for it to be 120 in the age of big monitors.